File: import-python-server/app/entities/product.py
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Optional
from ulid import ULID


PreciseAmount = int
Unit = int

class ImportStrategy(Enum):
	ATOMIC = "atomic"
	PARTIAL = "partial"

class CurrencyCode(Enum):
	"""
	Supported currency codes (ISO 4217 compliant, except BTC)
	"""
	USD = "USD"
	EUR = "EUR"
	JPY = "JPY"
	BRL = "BRL"
	BTC = "BTC"


class RoundingMode(Enum):
	TRUNCATE = "truncate"
	HALF_UP = "half_up"


@dataclass(frozen=True)
class PreciseNumber:
	"""
	Number represented as the smallest currency amount.
	"""
	amount: PreciseAmount
	unit: Unit

	def add(self, other: "PreciseNumber") -> "PreciseNumber":
		max_unit = max(self.unit, other.unit)
		scaled_self = self.amount * 10 ** (max_unit - self.unit)
		scaled_other = other.amount * 10 ** (max_unit - other.unit)
		return PreciseNumber(amount=scaled_self + scaled_other, unit=max_unit)
	
	def subtract(self, other: "PreciseNumber") -> "PreciseNumber":
		max_unit = max(self.unit, other.unit)
		scaled_self = self.amount * 10 ** (max_unit - self.unit)
		scaled_other = other.amount * 10 ** (max_unit - other.unit)
		return PreciseNumber(
			amount=scaled_self - scaled_other,
			unit=max_unit,
		)

	def multiply(self, other: "PreciseNumber") -> "PreciseNumber":
		return PreciseNumber(
			amount=self.amount * other.amount,
			unit=self.unit + other.unit,
		)

	def divide(self, other: "PreciseNumber", *, target_unit: int | None = None) -> "PreciseNumber | None":
		if other.amount == 0:
			return None

		if target_unit is None:
			target_unit = max(self.unit, other.unit)

		scale = 10 ** (target_unit + other.unit - self.unit)
		result_amount = (self.amount * scale) // other.amount

		return PreciseNumber(amount=result_amount, unit=target_unit)
	
	def normalize(self, target_unit: int, *, rounding: RoundingMode = RoundingMode.TRUNCATE) -> "PreciseNumber":
		if target_unit < 0:
			raise ValueError("target_unit must be >= 0")

		if target_unit == self.unit:
			return self

		# Increase precision
		if target_unit > self.unit:
			scale = 10 ** (target_unit - self.unit)
			return PreciseNumber(
				amount=self.amount * scale,
				unit=target_unit,
			)

		# Reduce precision — rounding required
		diff = self.unit - target_unit
		factor = 10 ** diff

		if rounding is RoundingMode.TRUNCATE:
			return PreciseNumber(
				amount=self.amount // factor,
				unit=target_unit,
			)

		if rounding is RoundingMode.HALF_UP:
			half = factor // 2
			adjustment = half if self.amount >= 0 else -half
			return PreciseNumber(
				amount=(self.amount + adjustment) // factor,
				unit=target_unit,
			)

		raise ValueError(f"Unsupported rounding mode: {rounding}")
	
	
	def __add__(self, other: "PreciseNumber") -> "PreciseNumber":
		return self.add(other)

	def __sub__(self, other: "PreciseNumber") -> "PreciseNumber":
		return self.subtract(other)

	def __mul__(self, other: "PreciseNumber") -> "PreciseNumber":
		return self.multiply(other)


@dataclass(frozen=True)
class CurrencyType:
	"""
	Currency value
	"""
	code: CurrencyCode
	amount: PreciseAmount


@dataclass(frozen=True)
class ProductPrice:
	"""
	Prices in supported currencies
	"""
	usd: CurrencyType
	eur: CurrencyType
	jpy: CurrencyType
	brl: CurrencyType
	btc: CurrencyType


@dataclass(frozen=True)
class ProductBatch:
	id: ULID
	created: datetime
	strategy: ImportStrategy
	filename: str

	@classmethod
	def new(cls, filename: str, strategy: ImportStrategy) -> "ProductBatch":
		return cls(
			id=ULID(),
			created=datetime.now(timezone.utc),
			strategy=strategy,
			filename=filename,
		)
	
	def to_dict(self):
		return {
			"id": str(self.id),
			"created": self.created.isoformat(),
			"strategy": self.strategy.value,
			"filename": self.filename,
		}


@dataclass(frozen=True)
class Product:
	id: ULID
	created: datetime
	batch_id: Optional[ULID]
	name: str
	prices: ProductPrice
	expiration: datetime

File: application-server/src/module/product/entity/Product.ts
import type { UlidID } from '@/types/entity';

export type ImportStrategy = 'atomic' | 'partial';

export interface ProductBatch {
	readonly id: UlidID;
	readonly created: Date;
	readonly strategy: ImportStrategy;
	readonly filename: string;
}

export interface Product {
	readonly id: UlidID;
	readonly created: Date;
	readonly batchID: UlidID | null;
	readonly name: string;
	readonly prices: ProductPrice;
	readonly expiration: Date;
}

export interface ProductPrice {
	/**
	 * US Dollar
	 */
	readonly usd: CurrencyType;
	/**
	 * Euro
	 */
	readonly eur: CurrencyType;
	/**
	 * Japanese Yen
	 */
	readonly jpy: CurrencyType;
	/**
	 * Brazilian Real
	 */
	readonly brl: CurrencyType;
	/**
	 * Bitcoin
	 */
	readonly btc: CurrencyType;
}

/**
 * Supported currency codes
 *
 * ISO 4217 compliant (except for BTC)
 */
export type CurrencyCode = 'USD' | 'EUR' | 'JPY' | 'BRL' | 'BTC';

/**
 * Precise amount
 *
 * Number represented as a `bitint` as the smallest currency amount.
 *
 * @example
 * ```
 * const usd: CurrencyAmount = 1090n // equivalent of $10.90
 * const btc: CurrencyAmount = 23405000n // equivalent of 0.23405000 BTC
 * ```
 */
export type PreciseAmount = bigint;

/**
 * Unit
 *
 * Number of decimals used to represent the smalles amount.
 */
export type Unit = number;

export interface PreciseNumber {
	amount: PreciseAmount;
	unit: Unit;
}

export interface CurrencyType {
	/**
	 * Currency code
	 */
	readonly code: CurrencyCode;
	/**
	 * Currency amount
	 */
	readonly amount: PreciseAmount;
}

export interface CurrencyDescriptor {
	/**
	 * Currency code
	 */
	readonly code: CurrencyCode;
	/**
	 * Unit
	 */
	readonly unit: Unit;
	/**
	 * Human-Readable currency name
	 */
	readonly description: string;
}

export const USD: CurrencyDescriptor = {
	code: 'USD',
	unit: 2,
	description: 'US Dollar',
};

export const EUR: CurrencyDescriptor = {
	code: 'EUR',
	unit: 2,
	description: 'Euro',
};

export const JPY: CurrencyDescriptor = {
	code: 'JPY',
	unit: 0,
	description: 'Japanese Yen',
};

export const BRL: CurrencyDescriptor = {
	code: 'BRL',
	unit: 2,
	description: 'Brazilian Real',
};

export const BTC: CurrencyDescriptor = {
	code: 'BTC',
	unit: 8,
	description: 'Bitcoin',
};

export function getCurrencyDescriptor(code: CurrencyCode): CurrencyDescriptor {
	switch (code) {
		case 'USD':
			return USD;
		case 'EUR':
			return EUR;
		case 'JPY':
			return JPY;
		case 'BRL':
			return BRL;
		case 'BTC':
			return BTC;
	}
}

export const CURRENCIES = [USD, EUR, JPY, BRL, BTC];
